### ISSUE TITLE
Require `prism` as a parser to analyse newly released ruby versions

### ISSUE THREAD

**Issue by Earlopain** [rubocop/rubocop#13617]
Context (don't want to spam it with rubocop specific stuff): https://github.com/whitequark/parser/issues/1046

The main `parser` maintainer will not backport new syntax changes to the parser gem anymore because:
* `prism` exists
* it's getting harder and harder to keep up with changes because CRuby itself is not using racc anymore for the old parser

I made a few PRs to import pending 3.4 changes and that was relativly straightforward, but only because ruby itself is not accepted new syntax for 3.4: https://bugs.ruby-lang.org/issues/20575. Still, there are some upstream changes still missing like `it` that made it in, and some other fringe cases which I don't believe anyone would actually notice or care about.

When 3.5 starts, there will probably more syntax changes, like the ones that were held back from 3.4.

Inevitably, RuboCop must do something in order to keep up. Is there someone here with the expertise to properly maintain the `parser` gem? I know it won't be me. I propose to simply only support the prism translation when analysing newer ruby versions, it will always be up-to-date and correct. The translation layer works quite well, there are barely any disabled tests and only a few prism-specific issues at the moment. Those can be worked out if needed.

@koic also had the same idea https://github.com/whitequark/parser/issues/1046#issuecomment-2505593801:

> It might be possible to design a transition where older Ruby versions continue to use the Parser gem, while newer Ruby versions rely on `Prism::Translation::Parser`, minimizing the impact on users.  

But I believe there is _at least_ one issue:

* `rubocop`  can't add prism as a dependency without bumping the major version.

Here's why I think so: https://github.com/rubocop/rubocop-ast/issues/282
When `rubocop-ast` added `prism` as its dependency, people complained because their environment lacks the necessary build tools to compile the native extension.

Maybe someone does step up to take care of `parser` but for now there's just a wish to keep it up-to-date. Let's talk about how to solve this now before it becomes a problem?

**Comment by bbatsov** [comment]
I think that's a reasonable plan overall, regardless of the short-term situation with `parser`. I think we will be able to push changes there ourselves if it comes to this. (or fully assume it's maintenance for a time)

> rubocop can't add prism as a dependency without bumping the major version.

I don't fully agree with this, as we've added plenty of new runtime deps along the way. To me adding `prism` won't be a big deal, if we don't have the same issues we did last time around with `rubocop-ast`. Perhaps those are already fixed?

**Comment by koic** [comment]
Here is a summary of my current thoughts. The Ruby syntax supported by the Parser gem and Prism differs as follows:

- **Parser gem**: Supports syntax for Ruby 2.0 through Ruby 3.3 (or potentially Ruby 3.4 and beyond, if supported).
- **`Prism::Translation::Parser`**: Supports syntax for Ruby 3.3 and later.

For example, the `it` syntax could potentially be supported by the Parser gem (either by myself or someone else). If `Prism::Translation::Parser` provides compatibility with the `it` syntax in the Parser gem, it would be considered supported.

It is unclear what syntax changes will come with Ruby 3.5, but for the time being, the Parser gem should remain the default parser. Many users have not yet migrated to Ruby 3.4 and are likely staying on older Ruby versions. This should be taken into account.

For this reason, the default should remain as the Parser gem for the time being. At this point, Prism should not be added as a dependency. Instead, it should be up to users to perform a `bundle update`.

> I propose to simply only support the prism translation when analysing newer ruby versions, it will always be up-to-date and correct. The translation layer works quite well, there are barely any disabled tests and only a few prism-specific issues at the moment. Those can be worked out if needed.

I cannot agree with this proposal at the current stage. There are still unresolved incompatibilities with `Prism::Translation::Parser` as far as RuboCop is concerned. Additionally, there are reports of syntactic issues in Prism’s core.

> Let's talk about how to solve this now before it becomes a problem?

While transitioning to `Prism::Translation::Parser` is likely to become a viable option in the future, making such a decision now is probably premature. We should observe how the situation evolves over time.

**To determine this, time is needed.**

As a side note, the change to the parser generator in ruby/ruby's `parser.y` involves moving from Bison to Lrama, not Racc. :-)

**Comment by Earlopain** [comment]
> To me adding prism won't be a big deal, if we don't have the same issues we did last time around with rubocop-ast. Perhaps those are already fixed?

There's nothing wrong with `prism`, it's just that it requires build tools and rubocop right now has no dependencies on any c extensions. So for people that don't have any c extensions through other gems right now, and are lacking build tools, their install would fail. I remember https://github.com/ruby-i18n/i18n/issues/685 for something similar.

I can understand both sides, its just a dependency but can cause breakage. Platform-specific prebuild gems of `prism` would probably help with that ([nokorigi](https://rubygems.org/gems/nokogiri) comes to mind as one example of doing that right now) but I have no idea how that works or how much effort that would be for `prism`.

In the end, I'm not sure how much pushback for the dependency there really would be, but I at least know it will not be zero.

---------

> It is unclear what syntax changes will come with Ruby 3.5, but for the time being, the Parser gem should remain the default parser.

And it should 100% stay that way, at least for versions `parser` properly supports. I'm not saying this should happen right now, but depending on how it goes with the parser gem it might need to happen for 3.5. Who knows how much Ruby will evolve until then.

For Ruby 3.4, it doesn't matter so much because it mostly made some fringe syntax that was previously valid now invalid, probably mostly found out because there was so much fresh parser work.

`it` also just works, although it's now accepting some syntax that it shouldn't. But I think that is much better than in the opposite direction where new syntax doesn't work.

> There are still unresolved incompatibilities with Prism::Translation::Parser as far as RuboCop is concerned

Yes, there are some. I will try to look into a few of those but I may be in over my head with that.

> Additionally, there are reports of syntactic issues in Prism’s core.

`prism` will be the default parser for ruby, so any issues it has will line up perfectly with how it really behaves (unless the user opts into `parse.y`). Why do you think that would be an issue?

> While transitioning to Prism::Translation::Parser is likely to become a viable option in the future, making such a decision now is probably premature. We should observe how the situation evolves over time.

Well, there are already some syntax changes that need to be ported. I'm just not familiar with parsers in general to attempt the bigger and more complex changes. The ones I already did barely touched the gramar.

> As a side note, the change to the parser generator in CRuby's parser.y involves moving from Bison to Lrama, not Racc :-)

Oops, yes. `parser` uses racc.

-----------

If someone does continue to maintain `parser`, then it is less pressing of course and I'll probably be able to chip away at the smaller things myself. But I haven't seen someone say, _yes, I'll continue the work_. **I'm under no circumstances** implying you should do that @koic, I just don't want this to come "out of nowhere" next year for the next ruby version so I want to talk about it now already and see what's possible.

**Comment by dvandersluis** [comment]
I think we should strongly consider forking (or transferring, if there's amenability there) parser to within @rubocop if new parsing changes are definitely no longer going to be done upstream. `itblock` (along the lines of `numblock`) is probably the only "missing" feature for Ruby 3.4 and we can get around it for now, but as new syntax features get added in 3.5 and above, it'll start to cause a problem quickly that we'll need to address, and IMO taking over maintenance of parser is drastically more feasible than rewriting rubocop to use prism without the translation layer.

It'd also be nice to be able to merge in some of the bug fixes that @Earlopain has created but have been sitting.

**Comment by Earlopain** [comment]
3.4 really was nicely lenient in what syntax changes it brought. That said, I would not spend much more time on `parser` myself. Instead I've been doing various compatibility work for the parser translator. `itblock` would be as simple as changing this here https://github.com/ruby/prism/blob/29f2a06f7bfa07fbb6b1f98aa52df0179336c325/lib/prism/translation/parser/compiler.rb#L1145-L1149 to be more like this https://github.com/ruby/prism/blob/29f2a06f7bfa07fbb6b1f98aa52df0179336c325/lib/prism/translation/parser/compiler.rb#L1360-L1364 (and some changes in the builder to pass it along correctly). Certainly easier than making changes to some arcane syntax file.

A change like https://bugs.ruby-lang.org/issues/20925 would probably be gotten for free simply because prism is behind it.

I do agree that the translator isn't where it needs to be right now, especially the issue with multibyte characters https://github.com/rubocop/rubocop/issues/13204 is bad.

But it's not so bleak. Out of all the 23442 test cases that RuboCop has, only 59 produce a mismatched ast. For tokens, it's not looking that great, 1167 don't match. But many occur from the same issue and others don't realistically matter (I think), like emitting `tLPAREN2` instead of `tLPAREN`. I would rather push to get these numbers down that to keep `parser` on life-support. Implementing complex ruby syntax changes in `parser` seems impossible, at least to me.

Who knows how long I will be interested in doing that and how much I can actually get done but I do think in the next 12 months until 3.5 many things can be accomplished. It's too early to say if it is a realistic goal, for now I'll just keep going along. Maybe you or someone else will help out as well :wink:

**Comment by koic** [comment]
Here are my current thoughts on these topics:

TL;DR

- The currently **available AST** suggests that `Prism::Translation::Parser` is likely to be a strong candidate for maintainability.  
- For RuboCop, the AST interface of the Parser gem is important.  
- Forking the Parser gem is probably not advisable.  

## `Prism::Translation::Parser` could become increasingly important

I think the key focus should be on the AST interface provided by the Parser gem. In other words, rather than directly working with Prism's AST, `Prism::Translation::Parser` becomes increasingly important. As already mentioned, `Prism::Translation::Parser` depends on the Parser gem's builder. However, instead of addressing issues across the entire Parser gem as an LALR parser, it allows for focused problem-solving within the builder. Depending on the Prism development team's approach, applying monkey patch (!? and other approaches) within `Prism::Translation::Parser` might also be an option.

## Forking the Parser gem Is probably not advisable

> I think we should strongly consider forking (or transferring, if there's amenability there) parser to within @rubocop if new parsing changes are definitely no longer going to be done upstream.  

At the least, I don't recommend forking the Parser gem. Ultimately, if it's feasible to maintain the gem within the RuboCop organization, it should also be possible to maintain it upstream, and contributions should be sent back upstream accordingly.

## AST node design for `it` block parameter syntax

Various designs can be considered for the AST representation of the `it` block parameter syntax, but I also think following the example of the `numblock` node and introducing an `itblock` node would be reasonable. This aligns with the existing AST design of the Parser gem, which RuboCop's AST users are already familiar with. In any case, this is a topic that should be discussed within the Parser gem repository.

## Thoughts on syntax tree

As a side note, the syntax tree that retains the most information is the CST (Concrete Syntax Tree). The level of abstraction that makes an AST easier to use depends on the user and their context. Since RuboCop already heavily depends on the Parser gem's AST, continuing to rely on it might be the most reasonable choice. While using Prism's raw AST could potentially improve performance, features like `on_send` callbacks and pattern macros such as `def_node_matcher` would also affect custom cop developers. Whether the large cost required to address such major incompatibilities is justified remains questionable.

## Miscellaneous

Ruby 3.4 has not introduced new syntax beyond the `it` syntax, and this decision is intentional. Matz has mentioned at various conferences that this is to ensure stability for Prism while Ruby 3.4 development. While it's unclear what syntax may appear in Ruby 3.5 and beyond, or how the Lrama development team is designing syntax trees, these discussions can be set aside until they materialize.

**Comment by Bo98** [comment]
Just to add 2 cents here from having to deal with environments where I cannot use native gems (though RuboCop doesn't really matter for my use case):

> So for people that don't have any c extensions through other gems right now, and are lacking build tools, their install would fail

It's worth noting that `prism` is a default gem which ships with Ruby. Though this would only work on Ruby 3.4 and later (assuming Prism 1.2.0 is even suitable here).

Such setup would also only provide backwards compatibility and not forwards compatibility. Running RuboCop under, for example, Ruby 3.4 with a `TargetRubyVersion` of 3.5 is probably uncommon, though it is presumably supported.

In terms of native extensions in general it's all gotten a bit more complicated now. RuboCop does have a native dependency specified, albeit a default gem: https://github.com/rubocop/rubocop/blob/0b4e1b4e16ad7673ada2593dc42695d8ac9c970b/rubocop.gemspec#L34 and `parser` has `racc` as a native dependency. For RubyGems (`gem install`), this is fine as it will use the default & bundled gems and skip these - and still does today.

Bundler behaves differently, but in Bundler versions shipped with Ruby 3.3 and earlier native installs can still be avoided as Bundler treats default gems special and can avoid installing them if you lock the versions appropriately (see example output below). This behaviour was however removed in recent Bundler so installing RuboCop via Bundler now requires a C compiler again under Ruby 3.4 (and recent 3.3.x but I'll need to test which).

Ruby's increasingly growing dependency on a C compiler toolchain is definitely one of my biggest gripes about the Ruby ecosystem in general, but that's a discussion for elsewhere.

<details>
<summary>Example output</summary>

Ruby 3.3.0:

```console
# gem install rubocop
Fetching unicode-emoji-4.0.4.gem
Fetching unicode-display_width-3.1.3.gem
Fetching ruby-progressbar-1.13.0.gem
Fetching ast-2.4.2.gem
Fetching parser-3.3.6.0.gem
Fetching rubocop-ast-1.37.0.gem
Fetching regexp_parser-2.10.0.gem
Fetching rainbow-3.1.1.gem
Fetching rubocop-1.69.2.gem
Fetching parallel-1.26.3.gem
Fetching language_server-protocol-3.17.0.3.gem
Successfully installed unicode-emoji-4.0.4
Successfully installed unicode-display_width-3.1.3
Successfully installed ruby-progressbar-1.13.0
Successfully installed ast-2.4.2
Successfully installed parser-3.3.6.0
Successfully installed rubocop-ast-1.37.0
Successfully installed regexp_parser-2.10.0
Successfully installed rainbow-3.1.1
Successfully installed parallel-1.26.3
Successfully installed language_server-protocol-3.17.0.3
Successfully installed rubocop-1.69.2
11 gems installed

# cd $(mktemp -d)
# echo "source 'https://rubygems.org'

gem 'rubocop'
" > Gemfile

# bundle install
Fetching gem metadata from https://rubygems.org/...........
Resolving dependencies...
Fetching json 2.9.1
Fetching racc 1.8.1
Installing json 2.9.1 with native extensions
Installing racc 1.8.1 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /usr/local/bundle/gems/json-2.9.1/ext/json/ext/generator
/usr/local/bin/ruby extconf.rb
checking for whether -std=c99 is accepted as CFLAGS... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

# echo "source 'https://rubygems.org'

gem 'rubocop'
gem 'racc', '=$(ruby -rracc -e 'puts Racc::VERSION')'
gem 'json', '=$(ruby -rjson -e 'puts JSON::VERSION')'
" > Gemfile

# bundle install
Fetching gem metadata from https://rubygems.org/...........
Resolving dependencies...
Bundle complete! 3 Gemfile dependencies, 14 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
```

Ruby 3.4.1:

```console
# gem install rubocop
Fetching unicode-emoji-4.0.4.gem
Fetching unicode-display_width-3.1.3.gem
Fetching ruby-progressbar-1.13.0.gem
Fetching ast-2.4.2.gem
Fetching parser-3.3.6.0.gem
Fetching rubocop-ast-1.37.0.gem
Fetching regexp_parser-2.10.0.gem
Fetching rainbow-3.1.1.gem
Fetching rubocop-1.69.2.gem
Fetching parallel-1.26.3.gem
Fetching language_server-protocol-3.17.0.3.gem
Successfully installed unicode-emoji-4.0.4
Successfully installed unicode-display_width-3.1.3
Successfully installed ruby-progressbar-1.13.0
Successfully installed ast-2.4.2
Successfully installed parser-3.3.6.0
Successfully installed rubocop-ast-1.37.0
Successfully installed regexp_parser-2.10.0
Successfully installed rainbow-3.1.1
Successfully installed parallel-1.26.3
Successfully installed language_server-protocol-3.17.0.3
Successfully installed rubocop-1.69.2
11 gems installed

# echo "source 'https://rubygems.org'

gem 'rubocop'
gem 'racc', '=$(ruby -rracc -e 'puts Racc::VERSION')'
gem 'json', '=$(ruby -rjson -e 'puts JSON::VERSION')'
" > Gemfile

# bundle install
Fetching gem metadata from https://rubygems.org/...........
Resolving dependencies...
Fetching json 2.9.1
Installing json 2.9.1 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /usr/local/bundle/gems/json-2.9.1/ext/json/ext/generator
/usr/local/bin/ruby extconf.rb
checking for whether -std=c99 is accepted as CFLAGS... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.
```
</details>

**Comment by Earlopain** [comment]
Interesting, thanks for your input. Do you have an idea why adding json as a dependency went without problems, even though it is a default gem? Some immediately started to notice when rubocop-ast did prism so I'm a bit confused. I don't think it has some crazy build requirements.

It does give me more confidence though.  `bigdecimal` is a native bundled gem that comes to mind and is not easily replaceable. Rails, among others, were forced to add it. They didn't have a native dependency before I believe.

With production systems, you can use groups to exclude installing RuboCop. RuboCop will always follow the released prism version closely, if not only to analyze new ruby versions like you said. It may also require  versions that exceed what is available by default in the latest ruby but on that I'm not so sure.

With the amount of bundled gems, you're bound to bump into some other dependency that's actually required at runtime though.

-----

Re bundler, native gems, and default versions. Have you seen this issue? https://github.com/rubygems/rubygems/issues/8180. It's specificy about concerns   TruffleRuby has but is also applicable to you.

**Comment by bbatsov** [comment]
> Forking the Parser gem Is probably not advisable

I don't think we'd ever get to forking Parser, regardless of what happens. We've always had a good relationship with its maintainers and if push comes to shove I was thinking of just taking over the maintenance of it, until we can switch fully to Prism. 

I also have one rather radical idea in mind, which will also be the simplest to implement - we drop Parser, bump RuboCop to 2.0 and reduce the minimum support Ruby version to 3.4. It will still require some work from our end, but at least it won't tie us to the maintenance of Parser. Still, doing this would be more painful for our users, so I'd rather avoid this if possible.

**Comment by Earlopain** [comment]
> I also have one rather radical idea in mind, which will also be the simplest to implement - we drop Parser, bump RuboCop to 2.0 and reduce the minimum support Ruby version to 3.4

That's rather bold, and also not needed I think. I would like to at least see non-EOL analysing.

`parser` is still required for the translation layer but theoretically the builder class could be inlined so that only `ast` remains. It would give some more flexibility on how the ast is constructed, without applying ugly monkeypatches for potential new node types, and also doesn't seem too bad to do if you consider `parser` a finished product now.

_But_ even with the translator you are still bound to the existing format. And when you are still on that format, there is no harm in continuing to use `parser` for versions that it understands.

**Comment by koic** [comment]
Yeah, rushing decisions should be avoided. Drastic incompatibilities could cause many users to feel disconnected and create a divide among RuboCop community, so they should be avoided. A way forward that minimizes user impact should be possible by considering both Parser gem and `Prism::Translation::Parser`.

**Comment by kddnewton** [comment]
Just wanted to pop in and say I'm happy to support whichever direction y'all want to go here. Let me know whatever is easiest for you/whichever direction you decide and I'm sure we can make it work. For what it's worth, @Earlopain has put in a ton of work on the translation layer just in the last week and I think it's getting much closer (@Earlopain do you have numbers?). I hope we can achieve 100% compatibility soon.

**Comment by Earlopain** [comment]
Sure, I can give some numbers. I use the following rake task to import rubocop test snippets:
<details><summary>Rake task</summary>
<p>

```rb
# frozen_string_literal: true

namespace :rubocop do
  desc "Ensure there's a local copy of rubocop"
  file "tmp/rubocop" do
    sh "git clone --depth=1 https://github.com/rubocop/rubocop tmp/rubocop"
  end

  desc "Ensure we have a fixtures directory for the rubocop tests"
  file "test/prism/fixtures/rubocop" do
    mkdir_p "test/prism/fixtures/rubocop"
  end

  desc "Import the rubocop tests"
  task import: ["tmp/rubocop", "test/prism/fixtures/rubocop"] do
    cp "tmp/rubocop/LICENSE.txt", "test/prism/fixtures/rubocop/LICENSE"

    require "bundler/inline"
    gemfile do
      source "https://rubygems.org"

      eval_gemfile("tmp/rubocop/Gemfile")
    end

    $LOAD_PATH.unshift File.expand_path("tmp/rubocop/spec")
    require_relative "../tmp/rubocop/spec/spec_helper"

    # RSpec.configure do |config|
    #   config.before :example do |example|
    #     $current_description = example.metadata[:full_description]
    #   end
    # end

    module Hook
      COLLECTED = []

      private

      def expect_offense(source, *, **replacements)
        test_file = caller_locations(1..1).first.absolute_path
        name = File.basename(test_file).sub("_spec.rb", "")

        code = format_offense(source, **replacements)
        code_no_annotations = RuboCop::RSpec::ExpectOffense::AnnotatedSource.parse(code).plain_source
        COLLECTED << code_no_annotations
        super
      end

      def expect_correction(code, ...)
        test_file = caller_locations(1..1).first.absolute_path
        name = File.basename(test_file).sub("_spec.rb", "")

        COLLECTED << code
        super
      end

      def expect_no_offenses(code, ...)
        test_file = caller_locations(1..1).first.absolute_path
        name = File.basename(test_file).sub("_spec.rb", "")

        COLLECTED << code
        super
      end
    end

    RuboCop::RSpec::ExpectOffense.prepend(Hook)

    RSpec::Core::Runner.run(Dir.glob('tmp/rubocop/spec/rubocop/cop/*/*'), STDERR, STDOUT)

    require "digest"
    Hook::COLLECTED.reject! { _1.encoding != Encoding::UTF_8 }
    Hook::COLLECTED.select! { Prism.parse_success?(_1, version: "3.3") }
    Hook::COLLECTED.each do |code|
      # codes.reject! { _1.encoding != Encoding::UTF_8 }
      # codes.select! { Prism.parse_success?(_1, version: "3.3") }

      name = Digest::MD5.hexdigest(code)
      filepath = "test/prism/fixtures/rubocop/#{name}.txt"
      File.write(filepath, code)
    end
  end

  desc "Clean up tmp files related to rubocop"
  task :clean do
    rm_rf "tmp/rubocop"
    rm_rf "test/prism/fixtures/rubocop"
    rm_rf "test/prism/snapshots/rubocop"
  end
end
```

</p>
</details> 

From the nature of those, it's mostly small snippets. But, only 42 ast differences on those and most related to something with regexp so it should be pretty easy to get that down even further with just a single change. At least two parser bugs are in there as well.

Tokens differ in 551 cases (less than half of what I started with last week), most still seem to be about differences with `tLPAREN2` vs `tLPAREN`. But there are also some string handling issues, particularly for heredocs with newlines and line continuations that I know can lead to actual issues.

All in all, there's now just one user-impacting issue remaing, one extra offense in `ruby-spec` that shouldn't be (https://github.com/ruby/prism/issues/2539). All the other pending reported issues against RuboCop (that I know of) will be fixed with the next prism release.

I guess the next step once it's translating all the tests would be to check it against whole codebases and see what pops up to make sure things don't just work by chance but overall it's looking pretty good I'd say.

---------

In the grand scheme of things I haven't really thought about next steps, mainly focusing on getting those numbers down. But there's still plenty of time to figure that out a bit later.

**Comment by bbatsov** [comment]
@Earlopain Great progress! I really appreciate all your work on this!

> Yeah, rushing decisions should be avoided. Drastic incompatibilities could cause many users to feel disconnected and create a divide among RuboCop community, so they should be avoided. A way forward that minimizes user impact should be possible by considering both Parser gem and Prism::Translation::Parser.

Of course. I'm the last person that will advocate for a radical course of action - just mentioning one of our options. I think it would be best for us to rely on `Prism::Translation::Parser` in the short-term and to devise a plan to switch to Prism's native AST format in the long-run. (mostly abstracted away by `rubocop-ast`)

**Comment by kddnewton** [comment]
In the long-run vision, I'd love it if we could eliminate the need for an intermediary AST layer. What I mean is that right now it goes `parser` -> `rubocop-ast` -> `rubocop`, and I'd prefer `prism` -> `rubocop`. If there are utilities that we need to add to the AST to make it easier to work with, let's just upstream them into prism so everyone can benefit. (Hopefully that also makes your maintenance burden lighter, because if the AST changes in a way that would break the utilities, it will be changed upstream first.)

**Comment by bbatsov** [comment]
@kddnewton Yeah, that would be great indeed. 

I guess the biggest reason for the existence of `rubocop-ast` was the pattern matching for AST nodes and some small extensions we made to the API of `parser` and `ast` to simplify our lives. I'm not sure if all of them will make sense in the context of Prism, but I'm pretty sure that the node pattern matching will be very useful there. Have you seen this https://docs.rubocop.org/rubocop-ast/node_pattern.html ?

**Comment by Earlopain** [comment]
Yeah, I don't think much needs to be upstreamed tbh. Prism already has that covered. I haven't spend too much time with the native ast format so I guess there may be something but it won't be too much. Though that can only convincingly be said after trying it out for real.

Node patterns are probably ok to stay in a gem? It's very useful when writing cops but probably they are a bit too specific. I guess it would be nice to have everything in one place but that maybe doesn't have to.

**Comment by Earlopain** [comment]
The ast is in a very good state now. I spend some more time fixing stuff up and would now consider it finished. It's able to accuratly translate all kinds of code now, even better than the parser gem does.

For example, the RuboCop codebase has one difference because `parser` doesn't handle `\\` in `<<~''` heredocs correctly. RuboCop tests themselves have differences because of regex escaping issues, nested heredocs, stuff like that. Even the gitlab codebase is now 1:1, sans one `parser` bug. I ran it over [real-world-ruby-apps](https://github.com/jeromedalbert/real-world-ruby-apps) and [real-world-rails](https://github.com/eliotsykes/real-world-rails) and pretty much only found more parser bugs.

<details><summary>Here is some of the interesting code I encountered where parser has trouble</summary>
<p>

```rb
# parser bug (?). if with then and no body has no begin location
if false then
elsif true
end

# parser bug (?). No begin location when first token is semicolon
if foo
  ;
end

# parser bug. dsym instead of sym for "" as hash key
{ "": 1 }

# parser bug. _1 in method inside block does not refer to the block variable
foo do
  def bar; _1 end
end

# parser bug. Tabs on second line mess up heredoc dedent
p(<<~ERROR_MESSAGE)
          foo
		bar
ERROR_MESSAGE

# parser bug. Null byte between `a` and `b` turns it into a dstring
"ab"

# parser bug. Percent string with literal newline as delimiter
# keeps the newline after the line continuation as string content
# -----
%
"foo" \
"bar"

# -----
```

</p>
</details> 

Tokens still need a bit more work and I doubt they will reach a similar level. The internal state of `parser` causes output changes where I don't think it will be feasable to accuratly mimic it. _But_ good enough should be doable. Already RuboCop behaves as expected in whatever I try, aligning the tokens would just make sure it's not by chance.

Much of RuboCop only concerns itself with the ast anyways, and many of the existing differences are irrelevant for it.

Many thanks to @kddnewton for how incredibly fast he was getting to my PRs!

**Comment by Earlopain** [comment]
I have this PR open against prism for a while now to make is possible to add new node types like `itblock` but haven't recieved any feedback yet https://github.com/ruby/prism/pull/3443

I would love for someone to take a look and tell me if that is going in the right direction. The PR description should contain all the context.

**Comment by qbantek** [comment]
Please consider https://github.com/whitequark/parser/pull/1069 to prioritize this issue. Thank you!

**Comment by koic** [comment]
Of course, the plan will reflect the reality of the Parser gem. RuboCop already has Prism support based on `ParserEngine: parser_prism`, and after some considerations, it will be addressed in the near future.

**Comment by Earlopain** [comment]
A new prism version got released which should contain all the necessary pieces from that side. 

Here's how I would like to continue:
* Teach `rubocop-ast` about the new prism builder (https://github.com/rubocop/rubocop-ast/pull/354)
* Teach `rubocop-ast` about `itblock`
* Handle `itblock` in `rubocop` and others, as it makes sense 

This can happen right now. Since there have been so many changes to the parser translator, I'd like to wait a bit until other changes happen, so that those that explicitly opted in to prism will test it out and report obvious issues if there are any. 

Analyzing Ruby 3.4 is still marked as experimental. But users are already on Ruby 3.4 and there is no warning or anything like that (for the better, I'd say), so things would potentially break when `rubocop` or `rubocop-ast` get updated, instead of when the user updates `TargetRubyVersion`. For that reason I'd like to take it a bit slower, even it is technically still experimental. 

After a short while, `prism` can be added as a hard dependency to `rubocop-ast`. It can then automatically use the prism translator when Ruby >= 3.4 is requested. In RuboCop, the `ParserEngine` option would still be around, but users are not expected to set this themselves. It would then be more of a transitionary option I guess, with not much value anymore.

Let me know how all that sounds, pr if something should happen differently.

---- 

Overall, not that much to do anymore, and most (if not all) now depends on changes in rubocop itself since I believe prism is good to go now. Except of course for inevitable translation incompatibilities, which I'm sure there will be some of (but I'm feeling pretty good about it).

**Comment by koic** [comment]
Overall, it looks good. I think this approach is the better option for now. First, the impact should be limited to users who have explicitly specified `ParserEngine: parser_prism`.  

Regarding the future, switching the default parser for Ruby 3.4 is a difficult decision, but given that there is no clear outlook on the progress of `whitequark/parser`, eventually setting `Prism::Translation::Parser` as the default seems to be the best choice for the future of the Ruby ecosystem.  

In summary:  
- For Ruby 3.3 and earlier, the default backend parser remains `whitequark/parser`.  
- For Ruby 3.4 and later, the default backend parser will be `Prism::Translation::Parser`.

As already proposed, the default switch should be done at a different timing.

**Comment by bbatsov** [comment]
@Earlopain I also like your plan, and I don't really see any viable alternatives at this point. Let's move forward with it.

**Comment by kddnewton** [comment]
@Earlopain I think the other step that would be forward-thinking would be to start to design/port `def_node_matcher` within Prism. I currently have `Prism::Pattern` (https://github.com/ruby/prism/blob/main/lib/prism/pattern.rb) which is somewhat like `def_node_matcher`, but I'd like to make it as easy as possible to port.

I also wonder if there's a way that we could take the input to `def_node_matcher` _currently_ and translate it into matching Prism trees directly. That might be quite difficult, but could be an interesting challenge.

Either way, in order to properly manipulate Prism ASTs in the very long term (assuming porting to the Prism AST is an _eventual_ goal) we will need this functionality.

In terms of the design itself, I'd love for it to be object-based as opposed to string-based. Both `def_node_matcher` and `Prism::Pattern` effectively are their own languages with their own parsers, which makes it difficult to statically analyze from an LSP. I'd love to make something that an LSP could help with.

**Comment by bbatsov** [comment]
> Either way, in order to properly manipulate Prism ASTs in the very long term (assuming porting to the Prism AST is an eventual goal) we will need this functionality.

That's the end goal (moving to Prism natively), but it will certainly take us a while to get there. (lots of work + we need to support users of older Ruby versions for a while longer, which is another reason to stick with the Parser translation for now) Unless AI coding assistants become a lot smarter very fast, that is. 😁 

> In terms of the design itself, I'd love for it to be object-based as opposed to string-based. Both def_node_matcher and Prism::Pattern effectively are their own languages with their own parsers, which makes it difficult to statically analyze from an LSP. I'd love to make something that an LSP could help with.

Agreed. Actually, I recall we tried to do something about back in the day but we didn't get very far.

**Comment by Earlopain** [comment]
> I think the other step that would be forward-thinking would be to start to design/port def_node_matcher within Prism

> In terms of the design itself, I'd love for it to be object-based as opposed to string-based.

I guess the first step would be to design something in rubocop-ast itself (`def_node_matcher_dsl` or whatever) to dogfood. I think it would be very hard to do this in prism first because using that in rubocop doesn't sound much different to switching to prism fully.

**Comment by Earlopain** [comment]
> I'm wondering if it would be possible to translate the matcher as well?

I guess it could be helpful when moving over to prism where you can simply leave your matchers as is while focusing on the other code around it? Is that why you are proposing this?

I have no idea how complicated that would be.

**Comment by Earlopain** [comment]
I've made a series of PRs that will finish this up. The current plan is that the next RuboCop version (1.75) will always use prism for Ruby 3.4+ without anything that the user must do. The main PRs for that are https://github.com/rubocop/rubocop-ast/pull/374 and https://github.com/rubocop/rubocop/pull/14025

I didn't expect it to happen so quickly but I'm not complaining. https://github.com/rubocop/rubocop/pull/14026 would be hard to do properly otherwise.

I'm quite happy to see this resolved soon and I want to again highligh how easy and pleasant it was to bring this over the finish-line in `ruby/prism` thanks to @kddnewton's help. This was quite the project for me (though in no way did I this alone! Having a good base to build upon, which @koic greatly helped with, was a huge help).

Using `prism` natively is still looming in the distance somewhere, but let's think about that separately 😅

**Comment by bbatsov** [comment]
Indeed - step by step we'll get where we need to be! Thanks for all of your hard work on this front! 🙇‍♂

**Comment by dvandersluis** [comment]
Great work @Earlopain! 🎉
